lib/utils/cli/cli.go

```diff
@@ -32,65 +32,158 @@ type CommandArgs struct {
+	// FlagsToRemove is a list of flags to omit from the resulting command.
 	FlagsToRemove []string
 }
 
 // Update returns new command line for the provided command taking into account
 // flags that need to be added or removed as configured.
 //
+// If the flag needs to be replaced (possibly to update the value), it needs to be
+// placed into both FlagsToAdd and FlagsToRemove.
+//
 // The resulting command line adheres to command line format accepted by systemd.
 // See https://www.freedesktop.org/software/systemd/man/systemd.service.html#Command%20lines for details
-func (a *CommandArgs) Update(command []string, flagsToAdd ...Flag) (args []string, err error) {
-	ctx, err := a.Parser.ParseArgs(command)
+func (r *CommandArgs) Update(command []string) (args []string, err error) {
+	ctx, err := r.Parser.ParseArgs(command)
 	if err != nil {
 		return nil, trace.Wrap(err, "failed to parse command: %v", command)
 	}
+	for _, el := range r.addRemoveFlags(ctx) {
+		args = append(args, el.Format()...)
+	}
+	outputCommand := ctx.SelectedCommand.FullCommand()
+	return append([]string{outputCommand}, args...), nil
+}
+
+func (r *CommandArgs) addRemoveFlags(ctx *kingpin.ParseContext) (elems []Flag) {
+	var args []Flag
 	seen := make(map[string]struct{})
 	for _, el := range ctx.Elements {
 		switch c := el.Clause.(type) {
 		case *kingpin.ArgClause:
-			args = append(args, strconv.Quote(*el.Value))
+			if utils.StringInSlice(r.FlagsToRemove, c.Model().Name) {
+				continue
+			}
+			seen[c.Model().Name] = struct{}{}
+			args = append(args, NewArg(c.Model().Name, *el.Value))
 		case *kingpin.FlagClause:
-			if utils.StringInSlice(a.FlagsToRemove, c.Model().Name) {
+			if utils.StringInSlice(r.FlagsToRemove, c.Model().Name) {
 				continue
 			}
-			if _, ok := c.Model().Value.(boolFlag); ok {
-				switch *el.Value {
-				case "true":
-					args = append(args, fmt.Sprint("--", c.Model().Name))
-				case "false":
-					args = append(args, fmt.Sprint("--no-", c.Model().Name))
-				}
+			seen[c.Model().Name] = struct{}{}
+			if _, ok := c.Model().Value.(boolCmdlineFlag); ok {
+				elems = append(elems, newBoolFlag(c.Model().Name, *el.Value))
 			} else {
-				args = append(args, fmt.Sprint("--", c.Model().Name), strconv.Quote(*el.Value))
+				elems = append(elems, NewFlag(c.Model().Name, *el.Value))
 			}
-			seen[c.Model().Name] = struct{}{}
 		}
 	}
-	for _, flag := range append(a.FlagsToAdd, flagsToAdd...) {
-		if _, ok := seen[flag.Name]; !ok {
-			args = append(args, fmt.Sprint("--", flag.Name), strconv.Quote(flag.Value))
+	for _, flag := range r.FlagsToAdd {
+		if arg, ok := flag.(arg); ok {
+			if _, exists := seen[arg.name]; !exists {
+				args = append(args, arg)
+			}
 		}
 	}
-	outputCommand := ctx.SelectedCommand.FullCommand()
-	return append([]string{outputCommand}, args...), nil
+	for _, flag := range r.FlagsToAdd {
+		if _, ok := flag.(arg); ok {
+			// Args have been filtered above
+			continue
+		}
+		if _, exists := seen[flag.Name()]; !exists {
```

Comment:
Could this have been done as a part of the above loop, without needing this second one? Something like:

```go
for _, flag := range r.FlagsToAdd {
  if arg, ok := flag.(arg); ok {
    // Collect positional arguments that aren't yet present in the resulting command-line.
    if _, exists := seen[arg.name]; !exists {
      args = append(args, arg)
    }
  } else {
    // Collect flags that aren't yet present in the resulting command-line.
    if _, exists := seen[flag.Name()]; !exists {
      elems = append(elems, flag)
    }
  }
}
```